### PR TITLE
Allow multiple options in ini_file

### PR DESCRIPTION
**Pull request by GheRivero**
Some ini files will allow an option to appear several times
with different values (e.g. systemd services files)

Add a new option, 'unique', defaulting to yes, to specifie if
a option can appears once or multiples times.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
ini_file

##### ANSIBLE VERSION
2.4.0
```

```

##### SUMMARY
Options in a ini file can appear multiple times in the same section. Until now, only one was allowed, deleting the rest.

[default]
path=/usr/bin
path=/usr/sbin

```

```

---
This change is [Reviewable](https://reviewable.io/reviews/ansible/ansible/19536)

### CODE REVIEW

**Comment by bcoca** [comment]
how does this interact with deletion and updates?

**Comment by GheRivero** [comment]
The behavior is exactly the same when unique=yes (default)
With unique: no, when updating, it simply adds a new option and checks that the (option, value) pair is not duplicated.
When deleting with unique=no, it deletes only the first option occurrence (same as with unique=yes)... but I'm thinking that maybe it should delete all the options occurrences.

**Comment by calfonso** [comment]
cc @jpmens are you up for reviewing and/or testing this change out and let us know so we can take action on this?

File: lib/ansible/modules/files/ini_file.py [hunk 1]
```diff
@@ -96,6 +96,14 @@
        - If set to 'no', the module will fail if the file does not already exist.
          By default it will create the file if it is missing.
      version_added: "2.2"
+  unique:
+     required: false
+     choices: [ "yes", "no" ]
+     default: "yes"
+     description:
+       - If set to 'no', the option could appears several times in the section with diferent values.
```
Comment (jpmens) [inline lib/ansible/modules/files/ini_file.py]: please change the world _could_ by _may_.

**Comment by jpmens** [comment]
Looks good to me. Small language nit in doc; if you'd please fix that.

File: lib/ansible/modules/files/ini_file.py (hunk 1 above)
Comment (GheRivero) [reply lib/ansible/modules/files/ini_file.py]: Done

**Comment by jpmens** [comment]
shipit

**Comment by Scriptkiddi** [comment]
hey I was wondering in which ansible version i can expect this feature?

**Comment by StoneISStephan** [comment]
When will this PR get merged?
I have multiple options with the same name, but different values, and they all must exist.
There are multiple valid values for the same option (they get fed to a selector).


Edit:

I tested this by manually merging the changes.
I see that I get unwanted behaviour when running the task again:

```
- name: Add PHP Memory Limits to panel.ini
  ini_file:
    path: "{{ plesk_panel_ini }}"
    section: 'php'
    option: "{{ item.option }}"
    value: "{{ item.value }}"
    state: present
    no_extra_spaces: yes
    create: yes
    unique: no
  with_items:
    - { option: 'settings.performance.memory_limit.values[]', value: '128M' }
    - { option: 'settings.performance.memory_limit.values[]', value: '256M' }
    - { option: 'settings.performance.memory_limit.values[]', value: '512M' }
    - { option: 'settings.performance.memory_limit.custom', value: 'false' }
  tags: plesk, plesk-configure, plesk-panel-ini
```


ansible-playbook output (-vv):


>  [php]
>  settings.general.open_basedir.default="none"
> +settings.performance.memory_limit.values[]=256M
>  settings.performance.memory_limit.values[]=128M
>  settings.performance.memory_limit.values[]=512M
> -settings.performance.memory_limit.values[]=256M
>  settings.performance.memory_limit.custom=false
> 



For my use case, this is not an issue, but I can imaging that you'd only want to have "state changed" when there was an actual change.

This might be fixable if values got written to the bottom of a section, instead of at the beginning.
The module would then not even create a change in the first place.



Edit2:

From what I see, the code does not really know where a section ends.

Would be nice if the code extracted the current section ini_lines
then we could use:

option_changed = newline not in section_ini_lines

in order not to change values that are already defined in all sections


[customer1]
allowed_values=128M
allowed_values=256M
[customer2]
allowed_values=128M
allowed_values=256M
allowed_values=512M


if we'd run a playbook that defines for customer1 the permissions that are already granted,
we'd like to have Ansible just return that no changes were needed.
In order words, the Ansible module should know that in section [customer1] these values are already set.
Instead of changing the above to:

[customer1]
allowed_values=256M
allowed_values=128M
[customer2]
allowed_values=128M
allowed_values=256M
allowed_values=512M


on the second call of ini_file module, as the module writes to <section_index + 1>
and then removes the last (duplicate) entry.

**Comment by StoneISStephan** [comment]
There is also a typo in the PR:

mgs = 'option added'

should be:

msg = 'option added'

File: lib/ansible/modules/files/ini_file.py [hunk 2]
```diff
@@ -227,16 +236,24 @@ def do_ini(module, filename, section=None, option=None, value=None,
                         option_changed = ini_lines[index] != newline
                         changed = changed or option_changed
                         if option_changed:
-                            msg = 'option changed'
-                        ini_lines[index] = newline
+                            if unique:
+                                msg = 'option changed'
+                                ini_lines[index] = newline
+                            else:
+                                mgs = 'option added'
```
Comment (StoneISStephan) [inline lib/ansible/modules/files/ini_file.py]: should be:

msg = 'option added'

**Comment by GheRivero** [comment]
closed my mistake...

**Pull request closed without merging**